### ISSUE TITLE
Non-deterministic ordering of union types in diagnostic messages

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#2844]
We currently see non-deterministic diagnostic message in ~20 different ecosystem projects.

One way to reproduce this is to run something like the following:
```bash
for i in {1..10}; do ty check --no-progress --output-format=concise | rg tests/test_tools.py:17:17; done
```
in the `rich` ecosystem project where I get the following output (scroll to the right):
```
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, str | Unknown]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, str | Unknown]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, Unknown | str]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, str | Unknown]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, str | Unknown]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, Unknown | str]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, Unknown | str]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, str | Unknown]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, Unknown | str]]`
tests/test_tools.py:17:17: error[invalid-argument-type] Argument to function `next` is incorrect: Expected `SupportsNext[Unknown]`, found `Iterable[tuple[bool, str | Unknown]]`
```

**Comment by sharkdp** [comment]
A few findings:
* This only happens in multi-threaded mode
* It seems to depend on the fact that "a lot" of files are being checked, *even if those files do not appear in the import tree of `test_tools.py`*! It seems like when I remove more and more files from the rich project, the "wrong" ordering becomes less likely…
* I was not able to reproduce this in single-threaded runs by permuting the order of checking the files.

Open questions:
* Is the [`source_order`](https://github.com/astral-sh/ruff/pull/21983) field potentially problematic in the context of non-deterministic checking-order of files?

**Comment by sharkdp** [comment]
Sharing a few more results after more hours of debugging:

* In pylox, I see an instance of this where we get either of these
   ```diff
   - Object of type `deque[None | Unknown]` is not assignable to attribute `fields` of type …
   + Object of type `deque[Unknown | None]` is not assignable to attribute `fields` of type …
   ```
* This type is created by the generics solver. It starts with a lower-bound constraint: `(Unknown | None) ≤ _T@deque ≤ object`. This constraint is split into two by distributing over the union [here](https://github.com/astral-sh/ruff/blob/962387ab345b00e7443bee5f7aa3e9f14ba63b38/crates/ty_python_semantic/src/types/constraints.rs#L530-L547). At this point, iterating over the union consistently yields the order `Unknown | None`.
* However, when we later pull these constraints out of the BDD, I get `(None ≤ _T@deque, source_order=2)` and `(Unknown ≤ _T@deque, source_order=2)` 95% of the time, resulting in the `None | Unknown` error message. But 5% of the time, I see `(Unknown ≤ _T@deque, source_order=1)` and `(None ≤ _T@deque, source_order=2)` with two different `source_order`s, resulting in the flipped ordering.
* [This TODO comment](https://github.com/astral-sh/ruff/blob/962387ab345b00e7443bee5f7aa3e9f14ba63b38/crates/ty_python_semantic/src/types/constraints.rs#L2206-L2208) looks relevant, but so far, I couldn't directly trace it to the root cause here

**Comment by sharkdp** [comment]
I think it has something to do with whether or not we have seen an `Unknown ≤ _T@deque` constraint before, possibly in a completely unrelated file.

**Comment by sharkdp** [comment]
I tried to reproduce this deterministically with a patched version of ty that checks files in the order that I pass them in (which is not the default behavior!), but no luck so far. Really hoping this is not a multithreading bug.

**Comment by sharkdp** [comment]
With the patch at the end of this script, I get the following debug output. The first output is what I get >95% of the time, the second output is what I get when I can finally reproduce it again:

`None | Unknown`:

```
UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

PATH:
  so=1    (str ≤ _T@deque)
  so=2    (LiteralString ≤ _T@deque)

PATH:
  so=2    (LiteralString ≤ _T@deque)

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

PATH:
  so=2    (None ≤ _T@deque)
  so=2    (Unknown ≤ _T@deque)

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

PATH:
  so=1    (Unknown ≤ _T@partial)

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

PATH:
  so=1    (_T@list ≤ _T0)
  so=1    (_T0 = *)

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

PATH:
  so=1    (Unknown ≤ _T@deque)

UNION LOWER: Unknown | int ≤ _T@deque ≤ object
  element: Unknown
  element: int

UNION LOWER: Unknown | int ≤ _T@deque ≤ object
  element: Unknown
  element: int

UNION LOWER: Unknown | int ≤ _T@deque ≤ object
  element: Unknown
  element: int

PATH:
  so=1    (Unknown ≤ _T@deque)
  so=2    (int ≤ _T@deque)

pylox/containers/array.py:11:9: error[invalid-assignment] Object of type `deque[None | Unknown]` is not assignable to attribute `fields` of type `dict[Unknown, Unknown]`
tests/arrays/test_array_methods.py:5:6: error[unresolved-import] Cannot resolve imported module `pylox.lox`
tests/arrays/test_empty_pop.py:5:6: error[unresolved-import] Cannot resolve imported module `pylox.lox`
tests/arrays/test_empty_popleft.py:5:6: error[unresolved-import] Cannot resolve imported module `pylox.lox`
```

`Unknown | None`:

```
  so=1    (Unknown ≤ _T@partial)

PATH:
  so=1    (str ≤ _T@deque)
  so=2    (LiteralString ≤ _T@deque)

PATH:
  so=2    (LiteralString ≤ _T@deque)

PATH:
  so=1    (_T@list ≤ _T0)
  so=1    (_T@list = *)

PATH:
  so=1    (Unknown ≤ _T@deque)

UNION LOWER: Unknown | int ≤ _T@deque ≤ object
  element: Unknown
  element: int

UNION LOWER: Unknown | int ≤ _T@deque ≤ object
  element: Unknown
  element: int

UNION LOWER: Unknown | int ≤ _T@deque ≤ object
  element: Unknown
  element: int

PATH:
  so=2    (int ≤ _T@deque)
  so=2    (Unknown ≤ _T@deque)

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

PATH:
  so=1    (Unknown ≤ _T@deque)
  so=2    (None ≤ _T@deque)

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

UNION LOWER: Unknown | None ≤ _T@deque ≤ object
  element: Unknown
  element: None

pylox/containers/array.py:11:9: error[invalid-assignment] Object of type `deque[Unknown | None]` is not assignable to attribute `fields` of type `dict[Unknown, Unknown]`
tests/arrays/test_array_methods.py:5:6: error[unresolved-import] Cannot resolve imported module `pylox.lox`
tests/arrays/test_empty_pop.py:5:6: error[unresolved-import] Cannot resolve imported module `pylox.lox`
tests/arrays/test_empty_popleft.py:5:6: error[unresolved-import] Cannot resolve imported module `pylox.lox`
```


<details><summary>Details</summary>
<p>

```diff
diff --git a/.cargo/config.toml b/.cargo/config.toml
index f7d8e616f8..80548491ac 100644
--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -8,3 +8,6 @@ benchmark = "bench -p ruff_benchmark --bench linter --bench formatter --"
 # See: https://github.com/astral-sh/ruff/issues/11503
 [target.'cfg(all(target_env="msvc", target_os = "windows"))']
 rustflags = ["-C", "target-feature=+crt-static"]
+
+[build]
+target-dir = "target"
diff --git a/crates/ty_python_semantic/src/types/constraints.rs b/crates/ty_python_semantic/src/types/constraints.rs
index 97f570a50e..942da2888f 100644
--- a/crates/ty_python_semantic/src/types/constraints.rs
+++ b/crates/ty_python_semantic/src/types/constraints.rs
@@ -536,6 +536,16 @@ impl<'db> ConstrainedTypeVar<'db> {
         //   T ≤ (¬α & ¬β) ⇔ (T ≤ ¬α) ∧ (T ≤ ¬β)
         //   (α | β) ≤ T   ⇔ (α ≤ T) ∧ (β ≤ T)
         if let Type::Union(lower_union) = lower {
+            eprintln!(
+                "UNION LOWER: {} ≤ {} ≤ {}",
+                lower.display(db),
+                Type::TypeVar(typevar).display(db),
+                upper.display(db)
+            );
+            for lower_element in lower_union.elements(db) {
+                eprintln!("  element: {}", lower_element.display(db));
+            }
+            eprintln!();
             let mut result = Node::AlwaysTrue;
             for lower_element in lower_union.elements(db) {
                 result = result.and_with_offset(
@@ -2374,6 +2384,11 @@ impl<'db> InteriorNode<'db> {
             Node::Interior(interior).for_each_path(db, |path| {
                 let mut path: Vec<_> = path.positive_constraints().collect();
                 path.sort_by_key(|(_, source_order)| *source_order);
+                eprintln!("PATH:");
+                for (c, so) in &path {
+                    eprintln!("  so={so:<4} {}", c.display(db));
+                }
+                eprintln!();
                 sorted_paths.push(path);
             });
             sorted_paths.sort_by(|path1, path2| {
```

</p>
</details>

**Comment by AlexWaygood** [comment]
You noted in https://github.com/astral-sh/ruff/pull/23373 that `union_or_intersection_elements_ordering()` appeared to produce a more stable result for the union `str | Unknown` than `source_order` does. @carljm noted that we'd originally introduced `source_order` precisely because `union_or_intersection_elements_ordering()` is not guaranteed to produce a stable ordering of union elements across different runs of ty (it was never designed to do that), only within a single run of ty.

However, something that's interesting here is that `str` and `Unknown` occupy different variants in our `Type` enum. `str` is represented using a `Type::NominalInstance(..)` variant; `Unknown` is represented using a `Type::Dynamic(..)` variant. It makes sense that `union_or_intersection_elements_ordering()` would produce a stable ordering between runs of ty for two types that occupy different variants; it only falls back to comparing types by Salsa ID if the two types occupy the _same_ `Type` variant.

This doesn't explain why `source_order` would produce an unstable ordering for these two types, however. I don't really know anything about how `source_order` works, only how `union_or_intersection_elements_ordering()` works.

**Comment by carljm** [comment]
I think we have a pretty good handle on this now and several PRs in-flight; going to consolidate our multiple open issues around non-determinism, closing this as dupe of #1670
